File: sol/solution/0001-0100/0008/Solution.ts
function myAtoi(s: string): number {
    let i = 0, sign = 1, result = 0;
    while (s[i] === ' ') i++;
    if (s[i] === '+' || s[i] === '-') sign = s[i++] === '-' ? -1 : 1;
    while (s[i] >= '0' && s[i] <= '9') {
        let digit = parseInt(s[i++]);
        if (result > (2**31 - 1 - digit) / 10)
            return sign === 1 ? 2**31 - 1 : -(2**31);
        result = result * 10 + digit;
    }
    return result * sign;
}

File: sol/solution/0001-0100/0008/Solution.cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0, sign = 1, n = s.size(), result = 0;
        while (i < n && s[i] == ' ') i++;
        if (i < n && (s[i] == '+' || s[i] == '-')) sign = (s[i++] == '-') ? -1 : 1;
        while (i < n && isdigit(s[i])) {
            int digit = s[i++] - '0';
            if (result > (INT_MAX - digit) / 10)
                return sign == 1 ? INT_MAX : INT_MIN;
            result = result * 10 + digit;
        }
        return result * sign;
    }
};